src/main/java/com/faforever/client/filter/FilterMultiCheckboxController.java

```diff
@@ -0,0 +1,177 @@
+    return property.getValue();
+  }
+
+  @Override
+  public MenuButton getRoot() {
+    return root;
+  }
+
+  public void setText(String text) {
+    this.text = text;
+    root.setText(text);
+  }
+
+  public void setItems(List<U> items) {
+    this.sourceList = items;
+
+    if (sourceList.size() > ITEM_AMOUNT_TO_ENABLE_SEARCH_BAR) {
+      convertedSourceList = convertToStringItems(items);
+      filteredList = new FilteredList<>(FXCollections.observableList(convertedSourceList), item -> true);
+      listView.setItems(filteredList);
+      addSearchBar();
+      addListenersWithSearch();
+    } else {
+      listView.setItems(FXCollections.observableList(items.stream().map(this::convertToString).toList()));
+      addListeners();
+    }
+  }
+
+  private void addListenersWithSearch() {
+    JavaFxUtil.addListener(searchTextField.textProperty(), observable ->
+        filteredList.setPredicate(item -> StringUtils.containsIgnoreCase(item, searchTextField.getText())));
+    JavaFxUtil.addListener(filteredList.predicateProperty(), observable -> restoreSelectedItems());
+    JavaFxUtil.addListener(listView.getCheckModel().getCheckedItems(), (ListChangeListener<String>) this::invalidated);
+    JavaFxUtil.bind(root.textProperty(), Bindings.createStringBinding(() -> i18n.get("filter.category", text,
+        String.join(", ", selectedStringItems)), selectedStringItems));
+  }
+
+  private void restoreSelectedItems() {
+    selectedStringItems.forEach(item -> listView.getCheckModel().check(item));
+  }
+
+  private void addListeners() {
+    JavaFxUtil.addListener(listView.getCheckModel().getCheckedIndices(),
+        (InvalidationListener) observable -> invalidated(listView.getCheckModel().getCheckedIndices()));
+    JavaFxUtil.bind(root.textProperty(), Bindings.createStringBinding(() -> i18n.get("filter.category", text,
+        String.join(", ", listView.getCheckModel().getCheckedItems())), listView.getCheckModel().getCheckedItems()));
+  }
+
+  private void addSearchBar() {
+    searchTextField = TextFields.createClearableTextField();
+    searchTextField.setPromptText(i18n.get("chat.searchPrompt"));
+    contentVBox.getChildren().add(0, searchTextField);
+  }
+
+  private void invalidated(Change<? extends String> change) {
+    if (change.next()) {
+      if (change.wasAdded()) {
+        String checkedItem = change.getAddedSubList().get(0);
+        if (!selectedStringItems.contains(checkedItem)) {
+          selectedStringItems.add(checkedItem);
+          selectedItems.add(sourceList.get(convertedSourceList.indexOf(checkedItem)));
+        }
+      } else if (change.wasRemoved()) {
+        String uncheckedItem = change.getRemoved().get(0);
+        if (selectedStringItems.contains(uncheckedItem)) {
+          selectedStringItems.remove(uncheckedItem);
+          selectedItems.remove(sourceList.get(convertedSourceList.indexOf(uncheckedItem)));
+        }
+      }
+    }
+  }
+
+  private void invalidated(List<Integer> selectedIndices) {
```

Comment:
```suggestion
  private void invalidate(List<Integer> selectedIndices) {
```
Visible for testing should only need package private